File: Year3/CA341/Assignment2/src - Functional/bst.hs
-- This code is adapted from a Lab Task submission for CA320. As such elements of this code have previously been
-- submitted as graded work by myself for CA320.

data Tree a
  = Null
  | Node a (Tree a) (Tree a)
  deriving (Read, Show)

insert :: Ord a => a -> Tree a -> Tree a
insert node Null = Node node Null Null
insert node (Node root right left)
  | node < root = Node root (insert node right) left
  | otherwise = Node root right (insert node left)

makeTree :: Ord a => [a] -> Tree a
makeTree [] = Null
makeTree (l : ls) = insert l (makeTree ls)

inOrder :: Tree a -> [a]
inOrder Null = []
inOrder (Node value right left) = inOrder right ++ [value] ++ inOrder left

postOrder :: Tree a -> [a]
postOrder Null = []
postOrder (Node value right left) = postOrder right ++ postOrder left ++ [value]

preOrder :: Tree a -> [a]
preOrder Null = []
preOrder (Node value right left) = [value] ++ preOrder right ++ preOrder left

search :: (Ord a) => Tree a -> a -> Bool
search Null _ = False
search (Node value left right) toFind
  | toFind== value = True
  | toFind< value = search left toFind
  | toFind > value = search right toFind

File: Year3/CA341/Assignment2/src - Logic/bst.pl
% sampleTree(tree(<VALUE>, tree(<VALUE>, tree(<VALUE>, nil, nil), nil), nil)).
% sampleEmptyTree(nil).

insert(ToInsert, nil, tree(ToInsert, nil, nil)).
insert(X, tree(Y, Left, Right), tree(Y, NL, Right)) :-
    X=<Y, !,
    insert(X, Left, NL).
insert(X, tree(Y, Left, Right), tree(Y, Left, NL)) :-
    insert(X, Right, NL).

search(tree(X, _, _), X) :- !.
search(tree(Curr, Left, _), ToFind) :-
    ToFind<Curr, !,
    search(Left, ToFind).
search(tree(_, _, Right), ToFind) :-
    search(Right, ToFind).

inOrder(nil, []).
inOrder(tree(X, Left, Right), IO) :-
    inOrder(Left, LLeft),
    inOrder(Right, LRight),
    append(LLeft, [X|LRight], IO).

preOrder(nil, []).
preOrder(tree(X, Left, Right), PO) :-
    preOrder(Left, LLeft),
    preOrder(Right, LRight),
    append([X|LLeft], LRight, PO).

postOrder(nil, []).
postOrder(tree(X, Left, Right), PO) :-
    postOrder(Left, LLeft),
    postOrder(Right, LRight),
    append(LLeft, LRight, RightE1),
    append(RightE1, [X], PO).


